core/module/const_source_location_spec.rb

```diff
@@ -0,0 +1,227 @@
+
+        ConstantSpecs::ModuleA::CS_CONST301 = :const301_2
+        ConstantSpecs::ModuleA.const_source_location(:CS_CONST301).should == [__FILE__, 17]
+
+        ConstantSpecs::ParentA::CS_CONST301 = :const301_3
+        ConstantSpecs::ParentA.const_source_location(:CS_CONST301).should == [__FILE__, 20]
+
+        ConstantSpecs::ContainerA::ChildA::CS_CONST301 = :const301_5
+        ConstantSpecs::ContainerA::ChildA.const_source_location(:CS_CONST301).should == [__FILE__, 23]
+      end
+
+      it "searches a path in a module included in the immediate class before the superclass" do
+        ConstantSpecs::ParentB::CS_CONST302 = :const302_1
+        ConstantSpecs::ModuleF::CS_CONST302 = :const302_2
+        ConstantSpecs::ContainerB::ChildB.const_source_location(:CS_CONST302).should == [__FILE__, 29]
+      end
+
+      it "searches a path in the superclass before a module included in the superclass" do
+        ConstantSpecs::ModuleE::CS_CONST303 = :const303_1
+        ConstantSpecs::ParentB::CS_CONST303 = :const303_2
+        ConstantSpecs::ContainerB::ChildB.const_source_location(:CS_CONST303).should == [__FILE__, 35]
+      end
+
+      it "searches a path in a module included in the superclass" do
+        ConstantSpecs::ModuleA::CS_CONST304 = :const304_1
+        ConstantSpecs::ModuleE::CS_CONST304 = :const304_2
+        ConstantSpecs::ContainerB::ChildB.const_source_location(:CS_CONST304).should == [__FILE__, 41]
+      end
+
+      it "searches a path in the superclass chain" do
+        ConstantSpecs::ModuleA::CS_CONST305 = :const305
+        ConstantSpecs::ContainerB::ChildB.const_source_location(:CS_CONST305).should == [__FILE__, 46]
+      end
+
+      it "returns path to a toplevel constant when the receiver is a Class" do
+        Object::CS_CONST306 = :const306
+        ConstantSpecs::ContainerB::ChildB.const_source_location(:CS_CONST306).should == [__FILE__, 51]
+      end
+
+      it "returns path to a toplevel constant when the receiver is a Module" do
+        Object::CS_CONST308 = :const308
+        ConstantSpecs.const_source_location(:CS_CONST308).should == [__FILE__, 56]
+        ConstantSpecs::ModuleA.const_source_location(:CS_CONST308).should == [__FILE__, 56]
+      end
+
+      it "returns path to the updated value of a constant" do
+        ConstantSpecs::ClassB::CS_CONST309 = :const309_1
+        ConstantSpecs::ClassB.const_source_location(:CS_CONST309).should == [__FILE__, 62]
+
+        -> {
+          ConstantSpecs::ClassB::CS_CONST309 = :const309_2
+        }.should complain(/already initialized constant/)
+
+        ConstantSpecs::ClassB.const_source_location(:CS_CONST309).should == [__FILE__, 66]
+      end
+    end
+
+    describe "with statically assigned constants" do
+      it "searches location path the immediate class or module first" do
+        ConstantSpecs::ClassA.const_source_location(:CS_CONST10).should == [@constants_fixture_path, 78]
```

Comment:
For these cases, it would be better to define some constants in that file like `CS_CONST10_LINE = __LINE__ - 1`.